components/script/dom/bindings/buffer_source.rs

```diff
@@ -226,27 +216,138 @@ where
             BufferSource::Float64Array(buffer) |
             BufferSource::DataView(buffer) |
             BufferSource::ArrayBuffer(buffer) |
-            BufferSource::Default(buffer) => {
+            BufferSource::Default(buffer) => unsafe {
+                IsDetachedArrayBufferObject(*buffer.handle())
+            },
+        }
+    }
+
+    pub fn viewed_buffer_array_byte_length(&self, cx: JSContext) -> usize {
+        match &self.buffer_source {
+            BufferSource::ArrayBuffer(buffer) => {
                 assert!(self.is_initialized());
                 let mut is_shared = false;
                 unsafe {
                     if JS_IsArrayBufferViewObject(*buffer.handle()) {
-                        // If it is an ArrayBuffer view, get the buffer using JS_GetArrayBufferViewBuffer
                         rooted!(in (*cx) let view_buffer =
                             JS_GetArrayBufferViewBuffer(*cx, buffer.handle(), &mut is_shared));
-                        // This buffer is always created unshared
                         debug_assert!(!is_shared);
-                        // Detach the ArrayBuffer
-                        DetachArrayBuffer(*cx, view_buffer.handle())
+                        return GetArrayBufferByteLength(*view_buffer.handle());
                     } else {
-                        // If it's not an ArrayBuffer view, Detach the buffer directly
-                        DetachArrayBuffer(*cx, Handle::from_raw(buffer.handle()))
+                        return GetArrayBufferByteLength(*buffer.handle());
                     }
                 }
             },
+            _ => unreachable!(
+                "viewed_buffer_array_byte_length should only be called on ArrayBufferView"
+            ),
         }
     }
 
+    pub fn byte_length(&self) -> usize {
+        match &self.buffer_source {
+            BufferSource::Int8Array(buffer) |
+            BufferSource::Int16Array(buffer) |
+            BufferSource::Int32Array(buffer) |
+            BufferSource::Uint8Array(buffer) |
+            BufferSource::Uint16Array(buffer) |
+            BufferSource::Uint32Array(buffer) |
+            BufferSource::Uint8ClampedArray(buffer) |
+            BufferSource::BigInt64Array(buffer) |
+            BufferSource::BigUint64Array(buffer) |
+            BufferSource::Float32Array(buffer) |
+            BufferSource::Float64Array(buffer) |
+            BufferSource::Default(buffer) |
+            BufferSource::DataView(buffer) => unsafe {
+                JS_GetArrayBufferViewByteLength(*buffer.handle())
+            },
+            BufferSource::ArrayBuffer(buffer) => unsafe {
+                GetArrayBufferByteLength(*buffer.handle())
+            },
```

Comment:
This match logic, unlike others, does make the correct assumption about whether we are dealing with a view or a buffer. We should do this everywhere.